tf_agents/agents/ddpg/critic_rnn_network.py

```diff
@@ -62,136 +72,97 @@ def __init__(self,
-    self._dynamic_unroll = dynamic_unroll_layer.DynamicUnroll(cell)
+    self._obs_encoder = obs_encoder
+    self._lstm_encoder = lstm_encoder
     self._output_layers = output_layers
 
   # TODO(kbanoop): Standardize argument names across different networks.
   def call(self, inputs, step_type, network_state=None):
+    outer_rank = nest_utils.get_outer_rank(inputs, self.input_tensor_spec)
+    batch_squash = utils.BatchSquash(outer_rank)
+
     observation, action = inputs
-    observation_spec, _ = self.input_tensor_spec
-    num_outer_dims = nest_utils.get_outer_rank(observation,
-                                               observation_spec)
-    if num_outer_dims not in (1, 2):
-      raise ValueError(
-          'Input observation must have a batch or batch x time outer shape.')
-
-    has_time_dim = num_outer_dims == 2
-    if not has_time_dim:
-      # Add a time dimension to the inputs.
-      observation = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1),
-                                          observation)
-      action = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1), action)
-      step_type = tf.nest.map_structure(lambda t: tf.expand_dims(t, 1),
-                                        step_type)
-
-    observation = tf.cast(tf.nest.flatten(observation)[0], tf.float32)
-    action = tf.cast(tf.nest.flatten(action)[0], tf.float32)
+    observation, _ = self._obs_encoder(
+        observation,
+        step_type=step_type,
+        network_state=network_state)
 
-    batch_squash = utils.BatchSquash(2)  # Squash B, and T dims.
-    observation = batch_squash.flatten(observation)  # [B, T, ...] -> [BxT, ...]
+    observation = batch_squash.flatten(observation)
+    action = tf.cast(tf.nest.flatten(action)[0], tf.float32)
     action = batch_squash.flatten(action)
 
-    for layer in self._observation_layers:
-      observation = layer(observation)
-
-    for layer in self._action_layers:
-      action = layer(action)
-
-    joint = tf.concat([observation, action], -1)
-    for layer in self._joint_layers:
-      joint = layer(joint)
-
-    joint = batch_squash.unflatten(joint)  # [B x T, ...] -> [B, T, ...]
-
-    with tf.name_scope('reset_mask'):
-      reset_mask = tf.equal(step_type, time_step.StepType.FIRST)
-    # Unroll over the time sequence.
-    joint, network_state = self._dynamic_unroll(
-        joint,
-        reset_mask,
-        initial_state=network_state)
-
-    output = batch_squash.flatten(joint)  # [B, T, ...] -> [B x T, ...]
+    output, network_state = self._lstm_encoder(
+        inputs=(observation, action),
+        step_type=step_type,
+        network_state=network_state)
 
     for layer in self._output_layers:
       output = layer(output)
 
     q_value = tf.reshape(output, [-1])
-    q_value = batch_squash.unflatten(q_value)  # [B x T, ...] -> [B, T, ...]
-    if not has_time_dim:
-      q_value = tf.squeeze(q_value, axis=1)
+    q_value = batch_squash.unflatten(q_value)
```

Comment:
add back the comment explaining how things are being unsquashed.

Reply:
added